File: docker/lua/framework/runtime/skynet-pb-codec.lua
local ____lualib = require("lualib_bundle")
local __TS__ParseInt = ____lualib.__TS__ParseInt
local __TS__ObjectEntries = ____lualib.__TS__ObjectEntries
local __TS__Class = ____lualib.__TS__Class
local Error = ____lualib.Error
local RangeError = ____lualib.RangeError
local ReferenceError = ____lualib.ReferenceError
local SyntaxError = ____lualib.SyntaxError
local TypeError = ____lualib.TypeError
local URIError = ____lualib.URIError
local __TS__New = ____lualib.__TS__New
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["15"] = 23,["16"] = 26,["17"] = 28,["18"] = 29,["19"] = 29,["20"] = 29,["21"] = 29,["22"] = 30,["23"] = 31,["26"] = 37,["27"] = 37,["28"] = 37,["29"] = 37,["30"] = 37,["31"] = 37,["32"] = 37,["33"] = 37,["34"] = 37,["35"] = 37,["36"] = 37,["37"] = 37,["38"] = 37,["39"] = 37,["40"] = 37,["41"] = 37,["42"] = 37,["43"] = 37,["44"] = 37,["45"] = 60,["46"] = 61,["47"] = 61,["48"] = 61,["49"] = 62,["51"] = 65,["52"] = 65,["53"] = 65,["55"] = 66,["56"] = 67,["57"] = 70,["58"] = 69,["59"] = 73,["60"] = 75,["61"] = 76,["64"] = 81,["65"] = 84,["66"] = 84,["67"] = 84,["68"] = 84,["69"] = 84,["70"] = 84,["71"] = 84,["72"] = 92,["73"] = 93,["74"] = 94,["75"] = 95,["76"] = 97,["77"] = 98,["78"] = 101,["79"] = 102,["80"] = 103,["82"] = 105,["85"] = 108,["88"] = 112,["89"] = 113,["90"] = 73,["91"] = 116,["92"] = 117,["94"] = 118,["98"] = 121,["99"] = 122,["101"] = 123,["105"] = 126,["106"] = 116,["107"] = 129,["108"] = 130,["110"] = 131,["114"] = 133,["115"] = 134,["117"] = 135,["121"] = 137,["122"] = 129,["123"] = 140,["124"] = 142,["125"] = 143,["126"] = 140,["127"] = 146,["128"] = 146,["129"] = 146,["131"] = 147,["133"] = 148,["137"] = 150,["138"] = 151,["139"] = 153,["140"] = 160,["141"] = 146,["142"] = 163,["143"] = 164,["145"] = 165,["149"] = 167,["150"] = 168,["151"] = 170,["153"] = 171,["154"] = 171,["155"] = 171,["156"] = 171,["160"] = 174,["161"] = 176,["162"] = 163});
local ____exports = {}
local skynet = _G.require("skynet")
local pb
local protoc
local hasPb = _G.pcall(function()
    pb = _G.require("pb")
    protoc = _G.require("protoc")
end)
if not hasPb then
    _G.print("[WARN] lua-protobuf not found, pb codec disabled")
end
--- 消息类型映射表
local MSG_ID_TO_NAME = {
    [100] = "gateway.HeartbeatRequest",
    [101] = "gateway.HeartbeatResponse",
    [102] = "gateway.ConnectRequest",
    [103] = "gateway.ConnectResponse",
    [104] = "gateway.DisconnectNotify",
    [200] = "login.LoginRequest",
    [201] = "login.LoginResponse",
    [202] = "login.LogoutRequest",
    [203] = "login.LogoutResponse",
    [204] = "login.ValidateTokenRequest",
    [205] = "login.ValidateTokenResponse",
    [206] = "login.GetOnlineCountRequest",
    [207] = "login.GetOnlineCountResponse",
    [300] = "game.EnterGameRequest",
    [301] = "game.EnterGameResponse",
    [302] = "game.LeaveGameRequest",
    [303] = "game.LeaveGameResponse"
}
local MSG_NAME_TO_ID = {}
for ____, ____value in ipairs(__TS__ObjectEntries(MSG_ID_TO_NAME)) do
    local idStr = ____value[1]
    local name = ____value[2]
    MSG_NAME_TO_ID[name] = __TS__ParseInt(idStr)
end
____exports.SkynetPbCodec = __TS__Class()
local SkynetPbCodec = ____exports.SkynetPbCodec
SkynetPbCodec.name = "SkynetPbCodec"
function SkynetPbCodec.prototype.____constructor(self)
    self.initialized = false
    self.protoRoot = ""
    self:initProto()
end
function SkynetPbCodec.prototype.initProto(self)
    if not hasPb then
        skynet.error("[SkynetPbCodec] Protobuf library not available, codec disabled")
        return
    end
    self.protoRoot = "./lua/protos"
    local protoFiles = {
        "common_pb.desc",
        "login_pb.desc",
        "game_pb.desc",
        "gateway_pb.desc",
        "message_id_pb.desc"
    }
    for ____, file in ipairs(protoFiles) do
        local filepath = (self.protoRoot .. "/") .. file
        local f = _G.io.open(filepath, "rb")
        if f then
            local readFile = _G.load("local f = ...; local data = f:read(\"*all\"); f:close(); return data")
            local data = readFile(f)
            local ok = pb.load(data)
            if ok then
                skynet.error("[SkynetPbCodec] Loaded " .. file)
            else
                skynet.error("[SkynetPbCodec] Failed to load " .. file)
            end
        else
            skynet.error("[SkynetPbCodec] Proto file not found: " .. filepath)
        end
    end
    self.initialized = hasPb
    skynet.error("[SkynetPbCodec] Initialized")
end
function SkynetPbCodec.prototype.encode(self, messageType, message)
    if not hasPb then
        error(
            __TS__New(Error, "[SkynetPbCodec] Protobuf not available"),
            0
        )
    end
    local encoded = pb.encode(messageType, message)
    if not encoded then
        error(
            __TS__New(Error, "[SkynetPbCodec] Failed to encode " .. messageType),
            0
        )
    end
    return encoded
end
function SkynetPbCodec.prototype.decode(self, messageType, data)
    if not hasPb then
        error(
            __TS__New(Error, "[SkynetPbCodec] Protobuf not available"),
            0
        )
    end
    local decoded = pb.decode(messageType, data)
    if not decoded then
        error(
            __TS__New(Error, "[SkynetPbCodec] Failed to decode " .. messageType),
            0
        )
    end
    return decoded
end
function SkynetPbCodec.prototype.create(self, messageType, init)
    local result = init or ({})
    return result
end
function SkynetPbCodec.prototype.pack(self, msgId, messageType, message, session)
    if session == nil then
        session = 0
    end
    if not hasPb then
        error(
            __TS__New(Error, "[SkynetPbCodec] Protobuf not available"),
            0
        )
    end
    local payload = self:encode(messageType, message)
    local timestamp = skynet.time()
    local packet = {msgId = msgId, session = session, data = payload, timestamp = timestamp}
    return self:encode("common.Packet", packet)
end
function SkynetPbCodec.prototype.unpack(self, data)
    if not hasPb then
        error(
            __TS__New(Error, "[SkynetPbCodec] Protobuf not available"),
            0
        )
    end
    local packet = self:decode("common.Packet", data)
    local messageType = MSG_ID_TO_NAME[packet.msgId]
    if not messageType then
        error(
            __TS__New(
                Error,
                "[SkynetPbCodec] Unknown msgId: " .. tostring(packet.msgId)
            ),
            0
        )
    end
    local message = self:decode(messageType, packet.data)
    return {msgId = packet.msgId, messageType = messageType, message = message, session = packet.session}
end
return ____exports

File: server/src/framework/runtime/skynet-pb-codec.ts
/**
 * Skynet Protocol Buffer 编解码器
 * 基于 lua-protobuf 库实现
 * 
 * 注意：此文件会被 TypeScriptToLua 编译为 Lua 代码
 * lua-protobuf API: https://github.com/starwing/lua-protobuf
 */

/** @noSelfInFile */

import { IPbCodec } from '../core/interfaces';

// @ts-ignore
const skynet = _G.require('skynet');

// 在 Lua 环境中动态加载 lua-protobuf（可选）
// @ts-ignore
let pb: any;
// @ts-ignore
let protoc: any;
// @ts-ignore
const hasPb = _G.pcall(() => { pb = _G.require('pb'); protoc = _G.require('protoc'); });
if (!hasPb) {
    // @ts-ignore
    _G.print('[WARN] lua-protobuf not found, pb codec disabled');
}

/**
 * 消息类型映射表
 */
const MSG_ID_TO_NAME: Record<number, string> = {
  // Gateway: 100-199
  100: 'gateway.HeartbeatRequest',
  101: 'gateway.HeartbeatResponse',
  102: 'gateway.ConnectRequest',
  103: 'gateway.ConnectResponse',
  104: 'gateway.DisconnectNotify',
  // Login: 200-299
  200: 'login.LoginRequest',
  201: 'login.LoginResponse',
  202: 'login.LogoutRequest',
  203: 'login.LogoutResponse',
  204: 'login.ValidateTokenRequest',
  205: 'login.ValidateTokenResponse',
  206: 'login.GetOnlineCountRequest',
  207: 'login.GetOnlineCountResponse',
  // Game: 300-399
  300: 'game.EnterGameRequest',
  301: 'game.EnterGameResponse',
  302: 'game.LeaveGameRequest',
  303: 'game.LeaveGameResponse',
};

const MSG_NAME_TO_ID: Record<string, number> = {};
for (const [idStr, name] of Object.entries(MSG_ID_TO_NAME)) {
  MSG_NAME_TO_ID[name] = parseInt(idStr as string);
}

export class SkynetPbCodec implements IPbCodec {
  private initialized = false;
  private protoRoot = '';

  constructor() {
    this.initProto();
  }

  private initProto(): void {
    // 如果没有 pb 库，跳过初始化
    if (!hasPb) {
      skynet.error('[SkynetPbCodec] Protobuf library not available, codec disabled');
      return;
    }

    // 获取 proto 文件目录（相对于 Skynet 工作目录）
    this.protoRoot = './lua/protos';

    // 加载所有 proto 描述文件
    const protoFiles = [
      'common_pb.desc',
      'login_pb.desc',
      'game_pb.desc',
      'gateway_pb.desc',
      'message_id_pb.desc',
    ];

    for (const file of protoFiles) {
      const filepath = `${this.protoRoot}/${file}`;
      // @ts-ignore
      const f = _G.io.open(filepath, 'rb');
      if (f) {
        // 使用 load 执行正确的 Lua 文件读取 (Lua 5.4)
        // @ts-ignore
        const readFile = _G.load('local f = ...; local data = f:read("*all"); f:close(); return data');
        // @ts-ignore
        const data = readFile(f);
        
        // 加载到 pb
        const ok = pb.load(data);
        if (ok) {
          skynet.error(`[SkynetPbCodec] Loaded ${file}`);
        } else {
          skynet.error(`[SkynetPbCodec] Failed to load ${file}`);
        }
      } else {
        skynet.error(`[SkynetPbCodec] Proto file not found: ${filepath}`);
      }
    }

    this.initialized = hasPb;
    skynet.error('[SkynetPbCodec] Initialized');
  }

  encode(messageType: string, message: any): Uint8Array {
    if (!hasPb) {
      throw new Error('[SkynetPbCodec] Protobuf not available');
    }
    // pb.encode 返回 Lua string，需要转换为 Uint8Array
    // @ts-ignore
    const encoded = pb.encode(messageType, message);
    if (!encoded) {
      throw new Error(`[SkynetPbCodec] Failed to encode ${messageType}`);
    }
    // 在 Lua 中，string 可以直接作为 bytes 使用
    return encoded as any;
  }

  decode(messageType: string, data: Uint8Array): any {
    if (!hasPb) {
      throw new Error('[SkynetPbCodec] Protobuf not available');
    }
    // @ts-ignore
    const decoded = pb.decode(messageType, data);
    if (!decoded) {
      throw new Error(`[SkynetPbCodec] Failed to decode ${messageType}`);
    }
    return decoded;
  }

  create(messageType: string, init?: any): any {
    // lua-protobuf 没有 create 方法，直接返回 init 或空表
    const result = init || {};
    return result;
  }

  pack(msgId: number, messageType: string, message: any, session: number = 0): Uint8Array {
    if (!hasPb) {
      throw new Error('[SkynetPbCodec] Protobuf not available');
    }
    const payload = this.encode(messageType, message);
    const timestamp = skynet.time();

    const packet = {
      msgId,
      session,
      data: payload,
      timestamp,
    };

    return this.encode('common.Packet', packet);
  }

  unpack(data: Uint8Array): { msgId: number; messageType: string; message: any; session: number } {
    if (!hasPb) {
      throw new Error('[SkynetPbCodec] Protobuf not available');
    }
    const packet = this.decode('common.Packet', data);
    const messageType = MSG_ID_TO_NAME[packet.msgId];

    if (!messageType) {
      throw new Error(`[SkynetPbCodec] Unknown msgId: ${packet.msgId}`);
    }

    const message = this.decode(messageType, packet.data);

    return {
      msgId: packet.msgId,
      messageType,
      message,
      session: packet.session,
    };
  }
}
